# Part I: Theory Questions



### k-Nearest Neighbor Classification

1. Assume that you have a large training dataset. Specify a disadvantage of the kNearest Neighbor method when using it during testing. State also your reason about your answer.


<font color=red>Answer: Larger datasets cost more time and memory than smaller ones during testing because at the testing process algoritm has to scan and store more data for prediction.This is a disadvange of kNearest Neighbor algoritm for large training datasets.</font> 

2. Create a 1-Dimensional classification dataset in which the 1-Nearest Neighbors method always gives a leave-one out cross validation error value of 1 (In other words, the method can’t guess correct class for a specific point in the dataset ). State also a proper explanation about your reasoning.


<font color=red>Answer: There can two possible reason for model's low accuracy,One is that 1 dimensional dataset's dimension size is too low to fit a model and this cause underfitting.Second is that k hyperparameter (K-nearest) is too low to predict true value.It must be increased a proper value for more accurate predictions.</font> 

3. Assume that you have the following training set of positive (+), negative (-) instances and a single test instance (o) in the figure below (Figure 1). Assume also that the Euclidean metric is used for measuring the distance between instances. Finally consider that every nearest neighbor instance affects the final vote equally.


• What is the class appointed to the test instance for K=1? State also reason behind your answer. 

<font color=red>Answer:  Appointed class is negative because the first nearest point is negative.</font> 

• What is the class appointed to the test instance for K=3? State also reason behind your answer.

<font color=red>Answer: Appointed class for 3 nearest neighbor is also negatif because all neighbor's weight are equal and negative neighbors outnumber.</font> 

• What is the class appointed to the test instance for K=5? State also reason behind your answer.

<font color=red>Answer: Appointed class for 3 nearest neighbor is positive because all neighbor's weight are equal and positive neighbors outnumber.</font> 

4. Fill the blanks with T (True) or F (False) for the statements below:

• If all instances of the data have the same scale then k-Nearest Neighbor’s performance increases drastically. (<font color=red>F</font> ) 

• While k-Nearest Neighbor performs well with a small number of input variables, it’s performance decreases when the number of inputs becomes large. (<font color=green>T</font>) 

• k-Nearest Neighbor makes no assumption about the functional form of the problem it handles. (<font color=green>T</font>)


### Linear Regression


1. Assume that you have five students have registered to a class and the class have a midterm and the final exam. You have obtained a set of their marks on two exams, which is in the table below:

You plan to a model which form’s is fθ(x) = θ0 + θ1x1 + θ2x2 for fitting the data above. The x1 shows midterm exam score while x2 shows square of the midterm score. Besides you plan to use feature scaling (using divide operation by the ”maxmin”, or range, of a feature) and mean normalization. What is the normalized value of the feature x(4) 2 ?

<font color=red>Answer: 0.38</font> 

2. Considering the figure below (Figure 2), which of the offsets used in linear regressions least square line fit? Assume that horizontal axis represents independent variable and vertical axis represents dependent variable. State your answer with your proper explanation.


<font color=red>Answer: We have to use vertical offsets for linear regressions least square.In this case vertical distance represent our residuals.</font> 

3. Considering the table below (Table 1), consisting of four training examples:
Assume that you are trying to fit the data above to the linear regression model fθ(x) = θ0 + θ1x1. Find the θ0 and θ1 values by using closed form solution (θ = (XTX)−1XTy). Also state dimension values of X, y and θ matrices. Finally show your calculations step by step.

<font color=red>Answer: In order to find thetas we have to do below process</font> 

In [25]:
#A^T*A*x=A^T*B
import numpy as np
aTranspose=np.array([1,2,4,0]).reshape(4,1)
aTa=np.matmul(np.array([1,2,4,0]),aTranspose)

In [27]:
#A^T*A*x
aTa
21*x1

array([21])

In [34]:
#A^T*B
b=np.array([0.5,1,2,0]).reshape(4,1)
aBa=np.matmul(np.array([1,2,4,0]),b)

In [35]:
aBa

array([10.5])

21x1=10.5

x1 is 0.5 and x0 is 0

4. State a valid reason for feature scaling and explain why it is a valid reason with respect to your reasoning.


<font color=red>Answer: In many machine learning algoritms we use Euclidian distance.If we don't normalize our dataset this formula can give misleading outcomes.Big values dominate small values and small features become meaningless eventhough they can be quite significant.</font> 

# PART II: Movie Recommendation System

Problem Definition:Creating a user-based movie rating prediction system by using KNN and weighted KNN algoritm.

In [39]:
import numpy as np
import os
import pandas as pd 
os.chdir(r"E:\Masaüstü\bbm409-assignment1") 

In [40]:
#importing needed data
mv=pd.read_csv("movies.csv").copy()
rt=pd.read_csv("ratings_train.csv").copy()
myTest=pd.read_csv("ratings_test_without_gt_new.csv")
myTest2=myTest[myTest["rating"]!=-1]#feature çıkartmak için convert edilmesi gereken test datası 
myTest=myTest[myTest["rating"]==-1]#predict edilmesi istenen test değerleri

In [43]:
###### combining movies.csv and ratings_train.csv with respect to movieıD
#combined data
new=pd.merge(rt, mv, on='movieId', how='inner')
new=pd.merge(rt, mv, on='movieId', how='inner')
new.head()


userId  movieId  rating   timestamp             title  \
0       1        1     4.0   964982703  Toy Story (1995)   
1       5        1     4.0   847434962  Toy Story (1995)   
2      15        1     2.5  1510577970  Toy Story (1995)   
3      17        1     4.5  1305696483  Toy Story (1995)   
4      18        1     3.5  1455209816  Toy Story (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1  Adventure|Animation|Children|Comedy|Fantasy  
2  Adventure|Animation|Children|Comedy|Fantasy  
3  Adventure|Animation|Children|Comedy|Fantasy  
4  Adventure|Animation|Children|Comedy|Fantasy

In [44]:
#this method creates feature matrix for training data, returns training Data's feature matrix and trainingDataUsers matrix 
#parameters are combined data and movies.csv. Use only for training data
def dataConverter(new,mv):
    
    new["isAd"]=[1 if "Adventure" in i else 0 for i in new["genres"]]
    new["isAn"]=[1 if "Animation" in i else 0 for i in new["genres"]]
    new["isCh"]=[1 if "Children" in i else 0 for i in new["genres"]]
    new["isCo"]=[1 if "Comedy" in i else 0 for i in new["genres"]]
    new["isFa"]=[1 if "Fantasy" in i else 0 for i in new["genres"]]
    new["isRo"]=[1 if "Romance" in i else 0 for i in new["genres"]]
    new["isHo"]=[1 if "Horror" in i else 0 for i in new["genres"]]
    new["isAc"]=[1 if "Action" in i else 0 for i in new["genres"]]
    new["isDr"]=[1 if "Drama" in i else 0 for i in new["genres"]]
    new["isCr"]=[1 if "Crime" in i else 0 for i in new["genres"]]
    new["isTh"]=[1 if "Thriller" in i else 0 for i in new["genres"]]
    new["isMy"]=[1 if "Mystery" in i else 0 for i in new["genres"]]
    new["isSc"]=[1 if "Sci-Fi" in i else 0 for i in new["genres"]]
    new["isWa"]=[1 if "War" in i else 0 for i in new["genres"]]
    new["isDo"]=[1 if "Documentary" in i else 0 for i in new["genres"]]
    new2=new[new.columns[6:]].copy()
    new2["userId"]=new["userId"]
    trainingData=new2.groupby('userId').sum()
    trainingDataUsers=trainingData.index.to_numpy()
    trainingData=trainingData.to_numpy()
    return trainingData,trainingDataUsers

In [46]:
#this method creates feature matrix for test data, returns test Data's feature matrix and test DataUsers matrix 
#parameters are convert edilmesi gereken test datası(myTest2) and movies.csv. Use only for training data
def testDataConverter(new1,mv):
    new=new1[new1["rating"]!=-1]
    new1=new1[new1["rating"]==-1]
    new=pd.merge(new, mv, on='movieId', how='inner')
    new["isAd"]=[1 if "Adventure" in i else 0 for i in new.genres]
    new["isAn"]=[1 if "Animation" in i else 0 for i in new.genres]
    new["isCh"]=[1 if "Children" in i else 0 for i in new.genres]
    new["isCo"]=[1 if "Comedy" in i else 0 for i in new.genres]
    new["isFa"]=[1 if "Fantasy" in i else 0 for i in new.genres]
    new["isRo"]=[1 if "Romance" in i else 0 for i in new.genres]
    new["isHo"]=[1 if "Horror" in i else 0 for i in new.genres]
    new["isAc"]=[1 if "Action" in i else 0 for i in new.genres]
    new["isDr"]=[1 if "Drama" in i else 0 for i in new.genres]
    new["isCr"]=[1 if "Crime" in i else 0 for i in new.genres]
    new["isTh"]=[1 if "Thriller" in i else 0 for i in new.genres]
    new["isMy"]=[1 if "Mystery" in i else 0 for i in new.genres]
    new["isSc"]=[1 if "Sci-Fi" in i else 0 for i in new.genres]
    new["isWa"]=[1 if "War" in i else 0 for i in new.genres]
    new["isDo"]=[1 if "Documentary" in i else 0 for i in new.genres]
    new2=new[new.columns[5:]].copy()
    new2["userId"]=new["userId"]
    testData=new2.groupby('userId').sum()
    userIDVector=testData.index.to_numpy()
    testData=testData.to_numpy()
    return testData,userIDVector
    

In [47]:
#this method create and return a dataframe thats colums are moviesIds and rows are usersIds.Parameter is combined data.
#This basicly our known target
def trainTargetConverter(new):
    dic={}
    for i in new.index:
        if(new["userId"][i]  not in dic.keys()):
            dic[new["userId"][i]]={new["movieId"][i]:new["rating"][i]}
        else:
              dic[new["userId"][i]].update({new["movieId"][i]:new["rating"][i]})


    df = pd.DataFrame(dic) 
    df=df.T
    df=df.apply(lambda row: row.fillna(row.mean()), axis=1)
    df=df.sort_index()
    return df


In [48]:
#Assigning train trainUser test testUser
trainingData,trainingDataUsers=dataConverter(new,mv)
testData,testDataUsers=testDataConverter(myTest2,mv)
trainTarget=trainTargetConverter(new)

In [49]:
#Fit and Knn methods

from numpy import linalg as LA
def fitMethod(train,test,kPara,trainTarget,trainingDataUsers,testDataUsers):
    #cosine based similarity used
    similarityMatrix=((np.matmul(train, test.T))/(LA.norm(train)*LA.norm(test)))
    return KNNMethod(similarityMatrix,trainTarget,kPara,trainingDataUsers,testDataUsers)


#This method returns a dataframe that colums are all movies in the movies.csv and rows are testusersId
#return predictedValues
def KNNMethod(similarityMatrix,trainTarget,kPara,trainingDataUsers,testDataUsers):
    sm=similarityMatrix.copy()
    sm=pd.DataFrame(sm,index=trainingDataUsers,columns=testDataUsers)
    predictValues=[]
    dic={}
    for i in sm.columns:
        col=i
        sm=sm.sort_values(by=col, ascending=False)
        values=sm.index[0:kPara]
        dic[i]={x:(sum(trainTarget.loc[values,x])/len(values)) for x in trainTarget.columns}

    predictValues=pd.DataFrame(dic).T
   
    return predictValues

In [52]:
#this method calculates MAE.Parameters ara trainTarget and predictedValues and only for known target
def testMethod(trainTarget,predictValues):
    sumValue=0
    for i in predictValues.index:
        for y in predictValues.columns:
            sumValue=abs((trainTarget[y][i])-(predictValues[y][i]))+sumValue
        
    MAE=(1/(len(predictValues)*len(trainTarget.columns)))*sumValue
    return MAE

In [68]:
#crossValidation method is only for TESTING, it cant't be used in unknown targets.It returns average of number of error
def crossValidation(kFold,trainingData,kPara,trainTarget,trainingDataUsers):#trainingData is data ,k fold
    
    data=trainingData.copy()
    data2=trainingDataUsers.copy()
    folds = np.array_split(data, kFold)
    foldsUsers=np.array_split(data2,kFold)
    listOfAccu=[]
    for i in range(kFold):
        tempTrain=folds.copy()
        tempTrainUser=foldsUsers.copy()
        test=folds[i].copy()
        testDataUsers=foldsUsers[i].copy()
        del tempTrain[i]
        del tempTrainUser[i]
        train=np.concatenate(tempTrain, axis=0).copy()
        trainingDataUsers=np.concatenate(tempTrainUser, axis=0).copy()
        predictValues=fitMethod(train,test,kPara,trainTarget,trainingDataUsers,testDataUsers)
        listOfAccu.append(testMethod(trainTarget,predictValues))
    print("MAE for k-fold ",kFold," ",sum(listOfAccu)/kFold)  
    return sum(listOfAccu)/kFold

In [54]:
#predict method takes userId and Movie tuple or list and predictedValues as parameters return and print userId and rating
def predictMethod(userAndMovie,predictedValues):
    listOfPrediction=[]
    for i in userAndMovie:
        print(i[0],",",predictedValues[i[1]][i[0]])
        listOfPrediction.append([i[0],predictedValues[i[1]][i[0]]])
    return listOfprediction 

In [61]:
#less test data is used in this cell due to the time issues. k parameter is 3
predicted=fitMethod(trainingData[20:],trainingData[0:20],3,trainTarget,trainingDataUsers[20:],trainingDataUsers[0:20])

In [51]:
predicted.head()

1         2         3         4        5         6         7       \
4   3.666667  2.833333  2.966319  3.144321  2.04735  3.500000  2.833333   
7   4.000000  2.833333  2.833333  2.960459  2.54735  3.449124  2.782457   
10  3.666667  2.833333  2.966319  3.144321  2.04735  3.500000  2.833333   
16  3.666667  2.833333  2.966319  3.144321  2.04735  3.500000  2.833333   
19  4.000000  2.833333  2.833333  2.960459  2.54735  3.449124  2.782457   

      8         9         10      ...    193565    193567    193571    193573  \
4   3.013669  2.763638  3.299652  ...  3.144321  3.144321  3.144321  3.144321   
7   2.829807  2.579776  3.500000  ...  2.960459  2.960459  2.960459  2.960459   
10  3.013669  2.763638  3.299652  ...  3.144321  3.144321  3.144321  3.144321   
16  3.013669  2.763638  3.299652  ...  3.144321  3.144321  3.144321  3.144321   
19  2.829807  2.579776  3.500000  ...  2.960459  2.960459  2.960459  2.960459   

      193579    193581    193583    193585    193587    193609  
4   3.144321  3.144321  3.144321  3.144321  3.144321  3.144321  
7   2.960459  2.960459  2.960459  2.960459  2.960459  2.960459  
10  3.144321  3.144321  3.144321  3.144321  3.144321  3.144321  
16  3.144321  3.144321  3.144321  3.144321  3.144321  3.144321  
19  2.960459  2.960459  2.960459  2.960459  2.960459  2.960459  

[5 rows x 8904 columns]

In [63]:
#MAE for k hyperparameter =3
MAE=testMethod(trainTarget,predicted)
print(MAE)

0.718745701427801


In [67]:
#trying different k parameter
for i in range(3,11):
    predicted=fitMethod(trainingData[20:],trainingData[0:20],i,trainTarget,trainingDataUsers[20:],trainingDataUsers[0:20])
    MAE=testMethod(trainTarget,predicted)
    print("MAE for ",i," hyperparameter",MAE)
    

MAE for  3  hyperparameter 0.718745701427801
MAE for  4  hyperparameter 0.7062708369633199
MAE for  5  hyperparameter 0.6368167278080952
MAE for  6  hyperparameter 0.6030906380597183
MAE for  7  hyperparameter 0.557272431223616
MAE for  8  hyperparameter 0.5363309773322302
MAE for  9  hyperparameter 0.5261050499817567
MAE for  10  hyperparameter 0.5329945908934768


In [69]:
#testing algoritm accuracy with cross validation kFold is 5 k parameter is 9.
#less training data is used in this test due to the time issue
predictedCv=crossValidation(5,trainingData[0:50],9,trainTarget,trainingDataUsers[0:50])

MAE for k-fold  5   0.4845092106479411


In [70]:
#for k=10 hyperparameter
predictedCv=crossValidation(5,trainingData[0:50],10,trainTarget,trainingDataUsers[0:50])

MAE for k-fold  5   0.48406924613027585


In [71]:
#weightedFitMethod and WeightedKnn method takes same parameters and return same values like normal fit and knn method.
#but in this knnMethod neighbors weight is like below
# (k is hyperparameter k) 1.neighbor*k+2.neighbor*k-1+....+k.neighbor*1
def weightedFitMethod(train,test,kPara,trainTarget,trainingDataUsers,testDataUsers):
    similarityMatrix=((np.matmul(train, test.T))/(LA.norm(train)*LA.norm(test)))
    return weightedKNNMethod(similarityMatrix,trainTarget,kPara,trainingDataUsers,testDataUsers)
    
def weightedKNNMethod(similarityMatrix,trainTarget,kPara,trainingDataUsers,testDataUsers):
    sm=similarityMatrix.copy()
    sm=pd.DataFrame(sm,index=trainingDataUsers,columns=testDataUsers)
    predictValues=[]
    dic={}
    kParaList=[k for k in range(kPara,0,-1)]
    sumKPara=sum(kParaList)
    for i in sm.columns:
        col=i
        sm=sm.sort_values(by=col, ascending=False)
        values=sm.index[0:kPara]
        mapped=zip(values,kParaList)
        mapped=[list(i) for i in mapped]
        dictionary={x[0]:x[1] for x in mapped}
        
        
        copyTrainTarget=trainTarget.loc[values,:].copy()
        
        copyTrainTarget=copyTrainTarget.mul(kParaList, axis=0)
            
        dic[i]={x:(sum(copyTrainTarget.loc[values,x])/sumKPara) for x in trainTarget.columns}
        
    
    predictValues=pd.DataFrame(dic).T
   
    return predictValues
    

In [72]:
for i in range(3,11):
    predictedWe=weightedFitMethod(trainingData[20:],trainingData[0:20],i,trainTarget,trainingDataUsers[20:],trainingDataUsers[0:20])
    MAE=testMethod(trainTarget,predictedWe)
    print("MAE for ",i," hyperparameter",MAE," in weighted KNN")

MAE for  3  hyperparameter 0.6990316897214891  in weighted KNN
MAE for  4  hyperparameter 0.6996309669354022  in weighted KNN
MAE for  5  hyperparameter 0.6771204180778738  in weighted KNN
MAE for  6  hyperparameter 0.6549371969429224  in weighted KNN
MAE for  7  hyperparameter 0.6292831653646285  in weighted KNN
MAE for  8  hyperparameter 0.6068984779981725  in weighted KNN
MAE for  9  hyperparameter 0.5895159628926648  in weighted KNN
MAE for  10  hyperparameter 0.5780313240983389  in weighted KNN


## Conclusion

According to test results the best k hyperparameter is 9 ,the best knn algoritm is simple knn (not weighted)